File: code/cpp/140.cpp
#include <cassert>
#include <iostream>
#include <map>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
  vector<string> wordBreak(string s, vector<string> &wordDict) {
    map<string, vector<string>> memo = {};
    return dp(s, wordDict, memo);
  }

  vector<string> dp(string s, vector<string> &wordDict,
                    map<string, vector<string>> &memo) {
    vector<string> ans;

    // find in memo
    if (memo.find(s) != memo.end())
      return memo.at(s);

    // s equal to ""
    if (!s.compare("")) {
      ans = {""};
      return ans;
    }

    for (int i = 0; i < wordDict.size(); i++) {
      // if s start with wordDict[i]
      if (s.rfind(wordDict[i], 0) == 0) {
        string reminder = s.substr(wordDict[i].size());
        vector<string> reminderResult = dp(reminder, wordDict, memo);

        vector<string> empty;
        if (reminderResult != empty) {
          vector<string> sResult;

          for (int j = 0; j < reminderResult.size(); j++) {
            string insertStr = reminderResult[j];
            // edge case
            if (reminderResult[j] != "") {
              insertStr.insert(0, " ");
            }
            insertStr.insert(0, wordDict[i]);
            sResult.push_back(insertStr);
          }

          // appending vector "sResult" to vector "ans"
          ans.insert(ans.end(), sResult.begin(), sResult.end());
        }
      }
    }

    memo.insert(make_pair(s, ans));
    return memo.at(s);
  }
};

int main() {
  Solution solution;
  vector<string> wordDict;
  vector<string> ans;
  vector<string> res;

  wordDict = {"cat", "cats", "and", "sand", "dog"};
  ans = {"cat sand dog", "cats and dog"};
  res = solution.wordBreak("catsanddog", wordDict);
  assert(res == ans);

  wordDict = {"apple", "pen", "applepen", "pine", "pineapple"};
  ans = {"pine apple pen apple", "pine applepen apple", "pineapple pen apple"};
  res = solution.wordBreak("pineapplepenapple", wordDict);
  assert(res == ans);

  wordDict = {"cats", "dog", "sand", "and", "cat"};
  ans = {};
  res = solution.wordBreak("catsandog", wordDict);
  assert(res == ans);
}

File: code/java/140.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;

class Solution {
    private StringBuilder builder = new StringBuilder();

    public List<String> wordBreak(String s, List<String> wordDict) {
        HashMap<String, List<String>> memo = new HashMap<>();
        return dp(s, wordDict, memo);
    }

    private List<String> dp(String s, List<String> wordDict, HashMap<String, List<String>> memo) {
        if (memo.containsKey(s))
            return memo.get(s);
        if (s.equals(""))
            return List.of("");

        List<String> ans = new ArrayList<>();

        for (String word : wordDict) {
            if (s.startsWith(word)) {
                String reminder = s.substring(word.length());
                List<String> reminderResults = dp(reminder, wordDict, memo);

                if (!reminderResults.equals(List.of())) {
                    List<String> sResults = reminderResults.stream().map(e -> resultAdd(e, word))
                            .collect(Collectors.toList());
                    ans.addAll(sResults);
                }
            } else {
                continue;
            }
        }

        memo.put(s, ans);
        return memo.get(s);
    }

    private String resultAdd(String e, String word) {
        // empty string builder
        builder.delete(0, builder.length());

        // edge case
        if (e.equals("")) {
            // Use StringBuilder inside a loop body when concatenating multiple strings
            return builder.append(word).append(e).toString();
        } else {
            return builder.append(word).append(" ").append(e).toString();
        }
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // [cat sand dog, cats and dog]
        System.out.println(solution.wordBreak("catsanddog", List.of("cat", "cats", "and", "sand", "dog")));

        // [pine apple pen apple, pine applepen apple, pineapple pen apple]
        System.out.println(
                solution.wordBreak("pineapplepenapple", List.of("apple", "pen", "applepen", "pine", "pineapple")));

        // []
        System.out.println(solution.wordBreak("catsandog", List.of("cat", "cats", "and", "sand", "dog")));
    }
}